pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -63,21 +87,58 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
-	// TODO: call pre-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
-	followup, err := next.Execute(msg)
+
+	// trigger the event if the message type is registered (if it's not outbound)
+	eventsTriggered := false
+	if !msg.Outbound {
+		eventsTriggered, err = s.sendEvent(msg, thid, next)
+		if err != nil {
+			return fmt.Errorf("send events failed: %w", err)
+		}
+	}
+
+	// if no events are trigger continue the execution
+	if !eventsTriggered {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// Register events on DID Exchange protocol.
+func (s *Service) Register(msgType string, ch chan<- event.DIDCommEvent) error {
```

Comment:
This function is currently conflating five activities:

1. Register a Go channel to listen to the event stream.
2. Register a Go channel to listen to workflow steps that require an action.
3. Set which events require an explicit callback action to progress the state machine.
4. Allow a different Go channel for each event type.
5. Filter the events that are returned to the client.

I argue that:
- (3) should have been done prior to the protocol starting.
- (4) seems unnecessary.
- (5) I could see some arguments for filtering... but I think the caller code looks the same even with filtering... and this is all local (not filtering a remote stream)...